### PR TITLE
Feature: Implemented MFA APIs

### PR DESCRIPTION
**Pull request by poovamraj**
### Changes

Implemented APIs to use MFA Login with OTP, OOB, Using recovery code and requesting MFA Challenge

New methods added to the public API - loginWithOTP, loginWithOOB, loginWithRecoveryCode, multifactorChallenge

Information on how to use this has been added to the README file as well

### Testing

I have added required unit tests and did manual testing to all the features as well. I have tested this with Android platform. Since this runs on JS, the iOS platform behaviour should be same. But never the less manual testing will be done for iOS as well

- [x] This change adds unit test coverage
- [x] This change has been tested on the latest version of the platform/language or why not

### Checklist

- [x] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
- [x] All existing and new tests complete without errors
- [x] All active GitHub checks have passed

### CODE REVIEW

File: src/auth/index.js [hunk 1]
```diff
@@ -350,6 +350,145 @@ export default class Auth {
...
+      .post('/oauth/token', {
+        ...payload,
+        client_id: this.clientId,
+        grant_type: 'http://auth0.com/oauth/grant-type/mfa-recovery-code',
+      })
+      .then(responseHandler);
+  }
+
+  /**
+   * Log in a user using the One Time Password code after they have received the 'mfa_required' error.
```
Comment (Widcket) [inline src/auth/index.js:457]: This documentation corresponds to the `loginWithOTP` method.

File: src/auth/index.js [hunk 2]
```diff
@@ -350,6 +350,145 @@ export default class Auth {
...
+   * @returns {Promise}
+   *
+   * @memberof Auth
+   */
+  multifactorChallenge(parameters = {}) {
+    const payload = apply(
+      {
+        parameters: {
+          mfaToken: {required: true, toName: 'mfa_token'},
+          challengeType: {required: true, toName: 'challenge_type'},
```
Comment (Widcket) [inline src/auth/index.js:478]: `challengeType` and `authenticatorId` are not required: https://auth0.com/docs/api/authentication#challenge-request

File: src/auth/__tests__/index.spec.js [hunk 3]
```diff
@@ -693,4 +693,240 @@ describe('auth', () => {
...
+      scope: 'openid profile email address phone',
+      tokenType: 'Bearer',
+    };
+
+    it('should require MFA Token and OOB Code', async () => {
+      expect.assertions(1);
+      expect(() => auth.loginWithOOB({})).toThrowErrorMatchingSnapshot();
+    });
+
+    it('binding code should be optional', async () => {
```
Comment (Widcket) [inline src/auth/__tests__/index.spec.js:800]: Could you please include a test case where the binding code is sent?

**Review by Widcket — CHANGES_REQUESTED** [review]
Looking good! The `multifactorChallenge` method has two required parameters that should be optional.

**Comment by poovamraj** [comment]
Thanks @Widcket that was a really nice catch. Will fix these and let you know!

File: src/auth/__tests__/index.spec.js [hunk 4]
```diff
@@ -693,4 +693,247 @@ describe('auth', () => {
...
+        auth.loginWithRecoveryCode(parameters),
+      ).resolves.toMatchSnapshot();
+    });
+  });
+
+  describe('Multifactor Challenge Flow', () => {
+    const parameters = {
+      mfaToken: '123',
+      challengeType: '123',
+      authenticatorId: '123',
```
Comment (Widcket) [inline src/auth/__tests__/index.spec.js:904]: Please add test cases that verify that `challengeType` and `authenticatorId` are optional.

Comment (Widcket) [reply src/auth/__tests__/index.spec.js:904]: Similar to what was done for `bindingCode` above.

Comment (poovamraj) [reply src/auth/__tests__/index.spec.js]: Added these changes 👍

**Review by Widcket — APPROVED** [review]

**Comment by cocojoe** [comment]
`I have added required unit tests and did manual testing to all the features as well. I have tested this with Android platform. Since this runs on JS, the iOS platform behaviour should be same. But never the less manual testing will be done for iOS as well`

This is vague, what versions exactly have been tested of each platform?

**Comment by poovamraj** [comment]
Currently I have manually tested the APIs in Android API Level 30 (Android 11). I have also tested the working in iOS with version 15.2. These testings were done with React Native version 66.4. I will test in the base version 65 as well

**Pull request merged**